components/script/dom/imagedata.rs

```diff
@@ -33,42 +34,54 @@ pub struct ImageData {
 }
 
 impl ImageData {
-    #[allow(unsafe_code)]
     pub fn new(
         global: &GlobalScope,
         width: u32,
         height: u32,
         mut data: Option<Vec<u8>>,
     ) -> Fallible<DomRoot<ImageData>> {
         let len = width * height * 4;
-        unsafe {
-            let cx = GlobalScope::get_cx();
-            rooted!(in (*cx) let mut js_object = ptr::null_mut::<JSObject>());
-            if let Some(ref mut d) = data {
-                d.resize(len as usize, 0);
-                let data = CreateWith::Slice(&d[..]);
-                Uint8ClampedArray::create(*cx, data, js_object.handle_mut()).unwrap();
-                Self::new_with_jsobject(global, None, width, Some(height), js_object.get())
-            } else {
-                Self::new_without_jsobject(global, None, width, height)
-            }
+
+        let cx = GlobalScope::get_cx();
+        rooted!(in (*cx) let mut js_object = ptr::null_mut::<JSObject>());
+        if let Some(ref mut d) = data {
+            d.resize(len as usize, 0);
+
+            let typed_array =
+                match create_typed_array::<ClampedU8>(cx, &d[..], js_object.handle_mut()) {
+                    Ok(typed_array) => typed_array,
+                    Err(_) => return Err(Error::JSFailed),
+                };
+
+            Self::new_with_data(
+                global,
+                None,
+                width,
+                Some(height),
+                CustomAutoRooterGuard::new(*cx, &mut CustomAutoRooter::new(typed_array)),
+            )
+        } else {
+            Self::new_without_data(global, None, width, height)
         }
     }
 
     #[allow(unsafe_code)]
-    fn new_with_jsobject(
+    fn new_with_data(
         global: &GlobalScope,
         proto: Option<HandleObject>,
         width: u32,
         opt_height: Option<u32>,
-        jsobject: *mut JSObject,
+        data: CustomAutoRooterGuard<Uint8ClampedArray>,
     ) -> Fallible<DomRoot<ImageData>> {
-        let heap_typed_array = match new_initialized_heap_typed_array::<ClampedU8>(
-            HeapTypedArrayInit::Object(jsobject),
-        ) {
-            Ok(heap_typed_array) => heap_typed_array,
-            Err(_) => return Err(Error::JSFailed),
-        };
+        let cx = GlobalScope::get_cx();
+        let heap_typed_array =
+            match new_initialized_heap_typed_array::<ClampedU8>(HeapTypedArrayInit::Data {
+                data: unsafe { data.as_slice() },
+                cx: cx,
+            }) {
+                Ok(heap_typed_array) => heap_typed_array,
+                Err(_) => return Err(Error::JSFailed),
```

Comment:
Either we are failing here where previously we wouldn't. Maybe try returning a type error here?

Or we are not producing a type error below at https://github.com/servo/servo/pull/31398/files#diff-74ee6b1aa263f228fb8648bb921e80b79654506e8b39570b764649a02f881402R89, don't know why. Maybe `get_buffer` is too liberal in returning an object even when it's of the wrong type?